### DISCUSSION TITLE
Say Hello! 👋

### DISCUSSION THREAD

**Discussion by carltongibson** [carltongibson/notes#29] [Being an open source maintainer]
Hi! 👋 I'm Carlton, former Django Fellow, and maintainer of many several things Django. 

The idea of this discussion group is to have a support and mentoring group for maintainers, and prospective maintainers, in the Django ecosystem. 

You're very welcome, so come and say hello. Introduce yourself… — and welcome aboard! ⛵ 

----

If you wanted to _Watch_ just the _Discussions_ on the repo (by going to the _Custom_ tab there) that might be an idea. 

----

Also, please do review the [code of conduct](https://github.com/carltongibson/notes/blob/main/README.md#code-of-conduct) if you've not done so before, or can't recall what it said. Thanks 🎁

**Comment by jefftriplett** [comment]
👋  Hello,

I'm Jeff (webology in online circles) and I help maintain [Django Packages](https://djangopackages.org). I'm interested in the discussion and whatever we can do with DP to help out.

**Reply by carltongibson** [reply]
Hi @jefftriplett — you win first to post 🏅 :) 

I know it's always implied work, but I think DP is an important component of the ecosystem, and there's opportunities there (slowly, slowly as always) to explore. 

Thanks for joining us.

**Reply by jefftriplett** [reply]
You hit the nail on the head with DP. We have done quite a bit of work to make the scoring more open and more helpful to make it easier to improve the quality of your package. The goal isn't to add a pointless checklist or get penalized like some of the commercial quality checkers do. Some of this is done and baked in, but we need some UI work before we launch it. 

My personal interests are to be in the loop so we can make DP better along with how to we create a better resource for the community. I also enjoyed mentoring during Djangonaut Spaces, and I learned a ton about how to do a better job of that going forward.

**Comment by nanorepublica** [comment]
Hey

I'm Andy (nanorepublica in most online circles) and I don't maintain anything of significance (...yet?) but have a general passion for the Django community so like to stick my nose into interesting discussion areas 😄 

I do have a couple of smaller package ideas that I need to finesse but they aren't quite high enough in the list of priorities these days.

**Reply by carltongibson** [reply]
Hey @nanorepublica — Welcome aboard! 😉

> …yet? 

That's a good part of the idea. I intend to write up some notes on the ins and out of that. The reason to do the discussion group here was to have it in the same place as I write those notes. So if anything particular comes to mind, do open a new discussion!

**Comment by tim-schilling** [comment]
Hi, I'm Tim :wave:

I help maintain a few packages ([django-debug-toolbar](https://github.com/jazzband/django-debug-toolbar/), [django-auth-adfs](https://github.com/snok/django-auth-adfs), [django-safemigrate](https://github.com/aspiredu/django-safemigrate/), [django-simple-history](https://github.com/jazzband/django-simple-history) though it's only to review @ddabble's work). Some of the things that have vexed me about package maintainership are:

- How do you get adoption of your project?
- How do you know when to agree to another person's architecture for a problem?
- What's the best way to ask someone, are you still interested in maintaining this package?
- How can I get feedback about the package?
- How can I get analytics on how my package is used?

Not looking for immediate answers, just fodder for the thought grinder. I'm looking forward to seeing how this develops!

**Reply by carltongibson** [reply]
Hey @tim-schilling — welcome aboard! 

Good topics, yes. 

I'll just speak to the "Are you still interested?" one... — I reckon the best way there is to phrase it as, "Would you like someone to ease your burden here?" — Many a maintainer will take you up on that. 🫠

**Comment by vossisboss** [comment]
Hi! I'm Meagen.

I am currently doing a lot of the community management work and promotion for the Wagtail community, but I am interested in learning how to do code reviews and more open source project management so that I can help relieve bottlenecks for our senior people.

For Wagtail, we have a mature project and I want to be doing more to help our senior devs. Some things I'm interested in discussing include:

- Project management strategies for connecting contributors with PRs and issues that they will succeed with
- The best ways to train newer reviewers or encourage them to give reviewing a go
- How to encourage new contributors to hang in there even if their code has been sitting in our queue for a long time
- Ways a less-experienced programmer like me can support our more experienced people with the skills I have now
- What skills I should be building to be a more effective intermediate-level contributor/reviewer

Looking forward to the discussions that come out of this!

**Reply by carltongibson** [reply]
Hi @vossisboss, welcome. We met at the Sprints at DjangoCon US in 2022 (it must have been). 

Those are good topics! Similar ones that we've battled with on Django. I have thoughts… but they may need gathering. (Happy to chat on any of them though in new discussions.)

**Comment by hvdklauw** [comment]
Hi, I'm Harro,

From the Netherlands, been doing Django since newforms, did webshops, communities, content management system, payment systems you name it I've build it. Also have a lot of experience peer reviewing code in many different codebases (both for merges as due diligence), so if you want something reviewed feel free to ask.

I don't maintain a lot of public stuff but if I find a bug I try to report it, fix it and get it merged. But usually I like to do some discussion on how to fix things and often I feel that package maintainers just ignore what I have to say unless I first make a PR, I guess I need help with just picking a solution instead of the best one. ;-)

Always leaned towards the DevOps/backend more than towards the frontend.

**Reply by carltongibson** [reply]
Hi @hvdklauw, welcome! 

I can say that maintainers often don't have enough time/energy for their projects, so it can be hard to engage with every thought offered. 

The flip side of that it that **most maintainers** are grateful for help when it falls in with the way they already manage their project. A question answered here, a PR reviewed there, and all of a sudden, you can find your other input given more of that limit time/energy. 

Does that make sense?

**Reply by sarahboyce** [reply]
(I am sorry I rambled a bit)

> But usually I like to do some discussion on how to fix things and often I feel that package maintainers just ignore what I have to say unless I first make a PR, I guess I need help with just picking a solution instead of the best one. ;-)

I'm guilty of this. 

Even though you are talking to a maintainer of the package, if you are talking about a specific problem which you have identified multiple solutions to, then you are in a position of "high context". 

Almost everyone else is in a position of "low context". 

This includes the maintainer.

Exceptions are when the maintainer has deep expertise in the problem space (Natalia calls this "seeing the matrix") or have touched this area recently (context switching hurts). But this is the exception, rather than the norm.

----

I think what you really want when you start the discussion is a "collaborator" - someone who will dive into the topic with you and perhaps be a co-committer of the contribution.

A "maintainer" can chose to collaborate with you, but it is a choice, and not really their role (at least for Django).

----

I wish I could "see the matrix" and engage with all discussions and say "what's best" - but I honestly don't know off the top of my head.

To roughly quote Carlton: "If you have spent a few hours diving into a Django ticket, looking into the code, past discussions etc - you are now the world expert of that ticket".

I also think reviewing PRs is an easier task than engaging in discussions.

If I am asked to proof read a presentation and suggest improvements  - I can do that, this is a well structured/defined task. 
If I'm asked which of two talk proposals on a complex topic, I'm vaguely familiar with, is going to lead to the better talk in my "expert" opinion - ooft that's tricky

**Reply by hvdklauw** [reply]
Thanks for the explanation (and terminology).

**Comment by tataraba** [comment]
Hi there. My name is Mario (usually PythonByNight in other online spaces).

I'm a toe-dipper when it comes to contributions to open source, though I am very much interested in "the plunge"... whatever that may mean. I've been interested in how exactly to connect those dots from curious onlooker to invested participator, not only for myself, but perhaps learning how to pull others along as well. I was recently part of the Djangonauts program, which was helpful and eye-opening. Would also appreciate a more focused approach, as this seems to be.

I consider myself a novice with Django itself, though it recently became part of my "full-time" job, so I'm sure I will gain much more familiarity. As far as maintaining a project, I have an idea for one, but not sure what it would take to get that over the edge.

Thanks for providing your time and everyone else who's adding to the discussion!

**Reply by carltongibson** [reply]
Hi Mario. Welcome aboard! ⛵ 

Yes... — that _OK, what's next?_ Is exactly what I had in mind, for prospective maintainers — so hopefully you're in the right place. 😜

**Comment by sarahboyce** [comment]
👋 Hello,

I'm Sarah and a [Django Fellow](https://www.djangoproject.com/fundraising/#fellowship-program) (so help maintain [Django core](https://github.com/django/django)). 

I have very similar interests to @vossisboss :grin: like...

* The ways to train/encourage reviewers
* How to encourage contributors to hang in there even if their code has been sitting in our queue for a long time, or getting multiple rounds of review

Then also...
* Encouraging a wider engagement on discussions
* Getting better analytics on how Django is used (stolen from @tim-schilling)
* Dealing with conflict or "heated" debates
* Reducing the lottery/burnout/retirement/bus factor
* Equity of opportunity/diversifying our leadership team and regular contributor pool
* (sounds selfish) Joyful maintenance - how to make sure our maintainers and regular contributors get positive feedback, know they are valued and appreciated, rather than "only being contacted when there's a problem or someone needs something from you"

**Reply by carltongibson** [reply]
Hey @sarahboyce 👋 

YYYEEESSS... all of those. OK, so there's already some topics we can start separate threads about. (I like how we're starting with the easy ones 😜)

**Reply by StephanieAG** [reply]
@sarahboyce fwiw, I don't think joyful maintenance sounds selfish at all; sounds to me more like a foundational pillar in a healthy community

besides, I'm sure I'm not the only relatively newer person around here who is constantly feeling surges of gratitude for a bunch of random people who openly discuss / work on a plethora of helpful and interesting stuff 😂 expressing gratitude and giving recognition in ways that feel meaningful for these people feels important

**Comment by ontowhee** [comment]
Hi! I'm interested in contributing more code and documentation, and I also want to participate in forum discussions and PR reviews. I'm not sure what particular package yet, but eventually I would like to gain enough knowledge to be a maintainer, and be able to guide others through the contribution process, too.

**Reply by carltongibson** [reply]
Hi @ontowhee — welcome aboard! ⛵

**Comment by VeldaKiara** [comment]
Hello 👋🏾, Velda Kiara, open source contributor and working towards becoming a maintainer for some of the OS projects I have contributed to like Django Debug Toolbar.

**Reply by carltongibson** [reply]
HI @VeldaKiara — sounds like you're in the right place. Welcome aboard. 🙂

**Comment by RosanaRufer** [comment]
Hi Carlton and everyone, pleasure to see y'all here! 
I've been a recent Djangonaut and planning to stick around, very fan of everyone here and the positive and humane people in the community.
Love 🫶

**Reply by carltongibson** [reply]
Hi @RosanaRufer — Welcome aboard! 

We're just getting started so a good time to join. 😊

**Comment by carltongibson** [comment]
Hi all — I'll try not to cross-post everywhere, but I don't know if we established a _How to get notified_... 🤔

Anyhow, just this time, I posted #30 as a conversation starter, based on some of the comments from last week. 

(Do post your own too!)

**Comment by mkalioby** [comment]
Hello, I'm Mohamed ElKalioby, I'm maintaining [django-mfa2](https://github.com/mkalioby/django-mfa2) and [django-passkeys](https://github.com/mkalioby/django-passkeys) and many other django packages.

My issue when I was bootstrap django-mfa2 was who to promote it so people start using it especially it was done so early, before webauthn API was even standardized.

**Reply by carltongibson** [reply]
Welcome aboard @mkalioby ⛵

**Comment by bckohan** [comment]
Hello! I'm [Brian Kohan](https://fosstodon.org/@bckohan). I've been using Django for work and personal projects since 2017. I've contributed the odd bug fix to packages I use here and there. I also maintain the following python packages: 

- [django-enum](https://github.com/bckohan/django-enum) (use almost any python Enum as a model field, also supports Flags)
- [enum-properties](https://github.com/bckohan/enum-properties) (A less verbose and generally more compatible alternative to dataclass enumerations)
- [django-typer](https://github.com/bckohan/django-typer) (use [Typer](https://github.com/fastapi/typer) (type hints) to define the CLI for your management commands)
- [django-render-static](https://github.com/bckohan/django-render-static) (use Django's template engines to render offline - includes js transpilers for enums and django.urls.reverse())
- [django-routines](https://github.com/bckohan/django-routines) (define named groups of management commands in settings that can be batch executed (e.g. ./manage.py routine deploy))
 - [sphinxcontrib-typer](https://github.com/sphinx-contrib/typer) (Generate command docs using [Typer's](https://github.com/fastapi/typer) rich formatting - [example here](https://igs-slm.readthedocs.io/en/latest/commands.html))
- [Site Log Manager](https://github.com/International-GNSS-Service/SLM) (a moderated web platform for managing GNSS ground station meta data implemented in Django)

Third party apps are definitely Django's secret sauce and are enabled by the scaffold the framework gives us more so than other "less opinionated" emerging options in Python. I think the continued health of the ecosystem is dependent on minimizing the time it takes for a new, well designed package that solves a common problem to be integrated into the community so that it is:

1. Easily discoverable
2. Has multiple *trusted* maintainers with publish privileges

Packages have always felt high risk, high reward for me. The more integral it is, the more potential upside to using it but also the more downside to its maintenance going poof. But also the kinds of things you might want to do to reduce risk in the ecosystem may also stifle the innovation that comes from its wild west nature. We're definitely reinventing too many wheels though. Lots to talk about!

**Reply by carltongibson** [reply]
Hi @bckohan 👋

> The more integral it is, the more potential upside to using it but also the more downside to its maintenance going poof. 

Yep 😅 — I became a collector of Django packages (all these years ago now) because I was building my business on them, and needed to **know** they'd work. The eternal challenge. 🤹

**Comment by SpudNiche** [comment]
Hey all! My name is Nic, and I've been using Django for 2 years now - last year I was exposed to the django community via "Django Chat" (thanks Carlton & Will!)

That said, I'm quite new to the game (I haven't contributed to any open source packages yet) but I'm eager to become familiar with the space and learn about contributing/maintaining. I'll likely lurk for a bit but hopefully start contributing something somewhere in the coming months. 🤞 

Thanks!

**Reply by carltongibson** [reply]
Hi @SpudNiche — Welcome aboard! Lurking is fine 🥳

**Comment by abe-101** [comment]
👋 Hello,
My name is Abe and I've been using Django for work and hobby projects, like [MergeCal](https://mergecal.org/), for about 3 years.

As I use Django, I'm interested in contributing back to the community. Recently, I've been contributing to [dj-stripe](https://github.com/dj-stripe/dj-stripe), but it's been challenging due to limited mentorship and a lack of context regarding the project's legacy.

Given that both of my hobby projects ([MergeCal](https://mergecal.org/) and [MyHebrewDates](https://myhebrewdates.com/)) are related to internet calendaring, I've developed an interest in contributing to the [icalendar](https://github.com/collective/icalendar) project.

Additionally, I recently initiated a Python package called [python-mergecal](https://github.com/mergecal/python-mergecal), which is designed to merge multiple calendars into one.

I'm particularly interested in finding mentorship opportunities to help me grow as a contributor and developer in these areas.

**Reply by carltongibson** [reply]
Hi @abe-101 — Welcome aboard!

**Reply by tim-schilling** [reply]
> it's been challenging due to limited mentorship and a lack of context regarding the project's legacy.

If you're not comfortable answering, don't feel obligated. But this is something I've thought about for a while. For the debug toolbar we created an [architecture doc](https://github.com/jazzband/django-debug-toolbar/blob/main/docs/architecture.rst). Even with that, I don't think it's enough to help someone come on. It seems like we may need a video or something that explains how the library gets used and how the logic is broken up.

Do you have any thoughts on what could be done to help you get up to speed on a new library?

**Reply by abe-101** [reply]
@tim-schilling the architecture doc is a really great starting point. As an outsider it give me a great overview of the project.
I wish every project did this!

**Comment by Stormheg** [comment]
Hello!

I'm Storm, and I build websites with [Wagtail](https://github.com/wagtail/wagtail) and Django during my day job. I'm a member of the Wagtail Core and accessibility teams since 2021ish. Django and Wagtail are amazingly useful pieces of software to me, and I believe the best way for me to give back is to step up and donate time to report, fix, and reply to discussions. We all benefit from open source.

Whilst I am a core team member, I don't often make contributions to Wagtail itself or review PRs. Wagtail is a pretty big project and I often feel a bit intimidated by it. As in: "I don't have enough knowledge about the code base and bigger picture to decide if this PR is heading in the right direction".

Instead I focus on smaller pieces. Maintaining useful Wagtail packages for example. There are not a lot of people who focus on this and I feel in my place doing that. It is honest work.

As of recently, I'm now an admin in the recently formed Django Commons organization for collectively maintaining Django packages. As an admin, I'm more in an overseeing role and help maintainers move their packages over to Django Commons and assist them in setting up automation for their project. Here's a quote from the Django Commons [README](https://github.com/django-commons/.github/blob/main/profile/README.md) about our goals:

> Django Commons has lofty goals for the future, but it will only be possible with your help:
>
>  - Compensate maintainers
>  - Identify and announce packages needing help
>  - Normalize and support maintainers periodically stepping back
>  - Support and encourage new contributors to step up

**Reply by carltongibson** [reply]
Hi @Stormheg 👋 — Welcome aboard! ⛵

**Reply by tim-schilling** [reply]
> Whilst I am a core team member, I don't often make contributions to Wagtail itself or review PRs. Wagtail is a pretty big project and I often feel a bit intimidated by it. As in: "I don't have enough knowledge about the code base and bigger picture to decide if this PR is heading in the right direction".

This 100%! Though I suppose the only way for us to do that is to start lurking on those PRs and dropping a code review every now and then.

**Reply by carltongibson** [reply]
The _Experienced Maintainters™_ often don't know what the answer is in a given case. Most times you have to dig-in quite hard to come up with an answer, same as anyone else. 

I feel like making sure that's well known would go a long way...

**Comment by nanuxbe** [comment]
Hi, I'm Emma and I am a serial project creator. I'm not really good at the maintenance part once I stop using the projects for myself though.

**Reply by carltongibson** [reply]
Hi @nanuxbe — welcome aboard! 


>  I'm not really good at the maintenance part once I stop using the projects for myself though.

Yes... common problem I reckon 😜 

I guess both signposting that that's the case for a project, and being open to handing them off (from almost day one) is all we can do there.

**Comment by AfiMaameDufie** [comment]
Hello 👋🏾, 

I am Afi, an open-source contributor interested in becoming a maintainer for OS projects.

**Reply by carltongibson** [reply]
Hi @AfiMaameDufie 👋 — Welcome aboard. Good to have you here.

**Comment by amirreza-sf80** [comment]
hi there
i'm amirreza
i maintain django-valkey and channels-valkey and have made some commits in some other projects

i'm new to open source (and programming in general) so i'm intereset in anything from the ground up

**Reply by carltongibson** [reply]
Hi @amirreza8002 — good to see you. 

One thing I thought you could do is blog about your packages, that would likely draw a link in Django News 📰

**Reply by amirreza-sf80** [reply]
hi
thank you 

I'll have to think on how a blog would look like🤔 I'm mostly standing on the shoulders of giants 
it's an interesting suggestion, hadn't thought of it
thanks 🙏

**Reply by tim-schilling** [reply]
Here are two angles you could take:

1. How did you go about creating django-valkey? I imagine that process was atypical of most libraries, but also very interesting as forks do happen.
2. What inspired you to create django-valkey, what are your immediate goals and what do you see the future of it as?

Bonus topic: how can Django better support valkey and redis?

**Reply by amirreza-sf80** [reply]
Hi!
thanks for the ideas
i started making myself a blog website.
I'll start writing (hopefully) after i manage to integrate `django-tailwind-cli` with the template I'm using.
and, well, translate all the hugo syntax to django (i copied the template from a github repo for hugo🥷)
that could be a subject as well 🤔
